Checks/PWR003/solution.c

```diff
@@ -0,0 +1,14 @@
+// PWR003: Explicitly declare pure functions
+
+// Depends only on its arguments
+// No side effects
+
+__attribute__((const)) double gravity();
```

Comment:
```suggestion
__attribute__((const)) extern double gravity();
```

Also, `__attribute__((const))` should be behind an `#ifdef` to prevent errors if the compiler does not support the attribute.

Reply:
omw

Reply:
Maybe @alvrogd has an opinion

Reply:
I was doing some research, what about
```cpp
#ifdef __GNU__
__attribute__((const))
#endif
double gravity();
```

Reply:
> Also, `__attribute__((const))` should be behind an `#ifdef` to prevent errors if the compiler does not support the attribute.

That's a good observation. There's no `#ifdef` mechanism in the code snippets in PWR003's entry just for simplicity. However, we should definitely enforce it for the benchmarks code, as we don't know which compilers will end up running them.



> I was doing some research, what about
> 
> ```c++
> #ifdef __GNUC__
> __attribute__((const))
> #endif
> double gravity();
> ```

The `__GNUC__` check should be enough. The PWR003 entry discusses GCC, clang, and ARM Compiler toolchain support for these extensions. `__GNUC__` is defined by both GCC and clang (because it supports GCC's extensions), and if I'm not mistaken, ARM's toolchain is based on GCC.

In any case, I would rather define it in line with what iñaki suggests:

```cpp
#ifdef __GNUC__
#define CONST_FUNCTION __attribute__((const))
#else
#define CONST_FUNCTION
#endif
```

It is a much more reusable construct, and better to stick to good practices.

Reply:
> icc?

Defines `__GNUC__` as well https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference/2024-0/additional-predefined-macros.html :)

Reply:
This is `icx`, `icx` is based on LLVM and so it makes sense it behaves just like `clang`. What about `icc`?

Reply:
My bad :grin: 

`icc` defines `__GNUC__` as well: https://www.intel.com/content/www/us/en/docs/cpp-compiler/developer-guide-reference/2021-10/additional-predefined-macros.html

Reply:
TSYM